134 api returns empty array when valid trips should exist

clukas1 opened the pull request:
Addressing issue raised, now there is more logic to convert gtfs transfers to raptor transfers with clear hierarchies.

- **Additional Transfers** generated by a Transfer Generator have lowest priority.
- **Child/Parent Transfers** are derived on a stop-stop basis in raptor. E.g. if parent stop holds transfers, stop A1 (child) will get all transfers from stop A (parent). See test cases to see all potential cases.
- **Explicit Transfers**: Even if parent transfers are defined in GTFS, if there is a direct transfer (e.g. A1-A2) is specified in GTFS, transfers derived from (A-A) will be overwritten.

In the process of implementation, I've also addressed a bug I found in our old transfer conversion. Where both additional transfers and gtfs transfers were all added (allowing duplicates), which in the end resulted in always using the shortest transfer, regardless where it was defined e062dc8.

munterfi reviewed (changes requested):
Thanks, @clukas1, really nice work!

I do have a few style and documentation suggestions.

Please improve the documentation, especially clarifying what "derive" transfers means in the context of parent and child stops. It would be helpful to describe the general idea, similar to what you outlined in the issue response, somewhere in the Javadoc of the method that encapsulates this logic.

Thanks!

clukas1 commented on src/main/java/ch/naviqore/raptor/router/RaptorRouterBuilder.java:
```diff
@@ -161,8 +164,9 @@ private StopContext buildStopContext(Lookup lookup) {
-            List<Transfer> currentTransfers = transfers.get(stopId);
-            int numberOfTransfers = currentTransfers == null ? 0 : currentTransfers.size();
+            Map<String, Transfer> currentTransfersMap = transfers.get(stopId);
+            Collection<Transfer> currentTransfers = currentTransfersMap == null ? new ArrayList<>() : currentTransfersMap.values();
```
Nice catch, 0e8a736

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverter.java:
```diff
@@ -87,27 +85,136 @@ private void addRoute(GtfsRoutePartitioner.SubRoute subRoute) {
+    private void processAllTransfers() {
+        addAdditionalTransfers();
+        processStopAndParentChildTransfers();
+        addGTFSTransfersWithPrecedence();
```
bc90441

clukas1 commented on src/test/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverterIT.java:
```diff
@@ -3,32 +3,303 @@
+    @Nested
+    class ManualSchedule {
+        /*
+        All tests run with a fixed set of stops and routes in a gtfs schedule as shown below:
```
bc90441

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverter.java:
```diff
@@ -87,27 +85,136 @@ private void addRoute(GtfsRoutePartitioner.SubRoute subRoute) {
+            if (stopTransfer.getTransferType() == TransferType.MINIMUM_TIME && stopTransfer.getMinTransferTime()
+                    .isPresent()) {
+                for (Stop toChildStop : stopTransfer.getToStop().getChildren()) {
+                    if (addedStops.contains(toChildStop.getId())) {
```
f9b38a9

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverter.java:
```diff
@@ -87,27 +85,136 @@ private void addRoute(GtfsRoutePartitioner.SubRoute subRoute) {
+     * @param fromStop the stop of interest, which transfers should be derived for
+     * @param toStop   the parent or child stop to process transfers for
+     */
+    private void deriveAllTransfersFromStop(Stop fromStop, Stop toStop) {
```
b9f0b08

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverter.java:
```diff
@@ -87,27 +85,136 @@ private void addRoute(GtfsRoutePartitioner.SubRoute subRoute) {
+     * @param parentStop the parent or child stop to collect transfers for
+     * @return a collection of transfers between the stops
+     */
+    private Collection<TransferGenerator.Transfer> collectTransfersBetweenStops(Stop fromStop, Stop parentStop) {
```
b9f0b08

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverter.java:
```diff
@@ -87,27 +85,136 @@ private void addRoute(GtfsRoutePartitioner.SubRoute subRoute) {
+    }
+
+    /**
+     * Adds direct transfers for a stop, ensuring minimum transfer time is handled correctly.
```
b9f0b08

clukas1 commented on src/test/java/ch/naviqore/service/gtfs/raptor/convert/GtfsToRaptorConverterIT.java:
```diff
@@ -3,32 +3,303 @@
+             * @throws NoSuchFieldException   If the field does not exist.
+             * @throws IllegalAccessException If the field is not accessible.
+             */
+            public static <T> T getPrivateField(Object object,
```
I completely agree that reflections are dark magic and should be used as little as possible. However, although ChatGPT agrees with you that this `ProgressTracker` pattern is best practice for such a case, I think I disagree. First of all by opening the class with this kind of dependency injection, why not just keep the getters public? In the end you're allowing every developer to access your class indirectly. Then I also don't like to add more complexity then needed to the production source code only for testing purposes. And lastly the thing I like about Reflections is, they are so obviously hacky that you really use them as a last resort and don't get inclined in overusing this kind of class access, which is different for the `ProgressTracker` where you kind of legitimize access to the private fields.

With this said, I'm happy you can live with my solution :)

clukas1 commented:
> Thanks, @clukas1, really nice work!
> 
> I do have a few style and documentation suggestions.
> 
> Please improve the documentation, especially clarifying what "derive" transfers means in the context of parent and child stops. It would be helpful to describe the general idea, similar to what you outlined in the issue response, somewhere in the Javadoc of the method that encapsulates this logic.
> 
> Thanks!

Thank you for your review @munterfi!

I've addressed all of your comments, especially re-writing the docstrings. A thing I've noticed while writing, is that we don't cover all potential cases (some parent/sibling transfers) but this might actually also be better this way and this is something the transfer generators are a good alternative for. To clarify some examples:

Stops: (note in this example parent stops also have departures)
A with Children A1, A2
B with Children B1, B2
C with Children C1, C2

Defined Transfers:
A-A
A1-A1
A-B
A1-C1

Resulting A1 transfers:
A1-A
A1-A1
A1-A2 (from A-A)
A-B
A1-B1 (from A-B)
A1-B2 (from A-B)
A1-C1

Not Covered but potentially possible transfers:
A1-C (from A1-C1)
A1-C2 (from A1-C2)

Hope you agree that adding more sibling parent logic would kind of be over interpreting transfers defined (or not defined) in GTFS.
